server/src/main/java/com/linecorp/centraldogma/server/internal/metadata/MetadataService.java

```diff
@@ -0,0 +1,853 @@
+    /**
+     * Removes {@link Permission}s from the specified {@code path}.
+     */
+    private CompletionStage<Revision> removeSpecificPermission(Author author, String projectName,
+                                                               JsonPointer path, String commitSummary) {
+        final Change<JsonNode> change = Change.ofJsonPatch(METADATA_JSON,
+                                                           new RemoveOperation(path).toJsonNode());
+        return metadataRepo.push(projectName, METADATA_REPO, author, commitSummary, change);
+    }
+
+    /**
+     * Replaces {@link Permission}s of the specified {@code path} with the specified {@code permission}.
+     */
+    private CompletionStage<Revision> replaceSpecificPermission(Author author,
+                                                                String projectName, JsonPointer path,
+                                                                Collection<Permission> permission,
+                                                                String commitSummary) {
+        final Change<JsonNode> change =
+                Change.ofJsonPatch(METADATA_JSON,
+                                   new ReplaceOperation(path, Jackson.valueToTree(permission)).toJsonNode());
+        return metadataRepo.push(projectName, METADATA_REPO, author, commitSummary, change);
+    }
+
+    /**
+     * Finds a collection of {@link Permission}s which belongs to the specified {@link User} from the specified
+     * {@code projectName} and {@code repoName}.
+     */
+    public CompletionStage<Collection<Permission>> findPermission(String projectName, String repoName,
+                                                                  User user) {
+        requireNonNull(projectName, "projectName");
+        requireNonNull(repoName, "repoName");
+        requireNonNull(user, "user");
+
+        if (user.isAdmin()) {
+            return CompletableFuture.completedFuture(RolePermission.ALL_PERMISSION);
+        }
+        return getProject(projectName).thenApply(metadata -> {
+            final Repo repo = metadata.repo(repoName);
+            final Member member = metadata.memberOrDefault(user.id(), null);
+
+            // If the member is guest.
+            if (member == null) {
+                return repo.rolePermission().guest();
+            }
+            final Collection<Permission> p = repo.userPermission().get(member.id());
+            if (p != null) {
+                return p;
+            }
+            return findPermissionByRole(repo, member.role());
+        });
+    }
+
+    /**
+     * Finds a collection of {@link Permission}s which belongs to the token of the specified
+     * {@link UserWithToken} from the specified {@code projectName} and {@code repoName}.
+     */
+    public CompletionStage<Collection<Permission>> findPermission(String projectName, String repoName,
+                                                                  UserWithToken userWithToken) {
```

Comment:
This method could be merged into `findPermission(..., User)` with an `instanceof` check given `UserWithToken` is a subtype of `User`.